### PR TITLE
[1015] Add trainee start date workflow

### PR DESCRIPTION
**Pull request by jhenderson**
### Context
https://trello.com/c/vF1XOj65/1015-l-add-commencement-date

### Changes proposed in this pull request
- Add commencement_date field to the Trainee
- Replace "Trainee ID" with "Trainee start date and ID" on the review draft page
- New training details page to capture both trainee ID and start date, including validation and progress tracking
- Add edit page for Trainee start date
- Non-draft will have separate confirmation for each field.
- Send commencement_date to DTTP
- Remove trainee ID field from new record page

### Guidance to review
Visit a draft and non-draft record play around with changing the trainee ID and start date

### CODE REVIEW

**Comment by jhenderson** [comment]
[[L] Add commencement date](https://trello.com/c/vF1XOj65/1015-l-add-commencement-date)

**Review by defong — CHANGES_REQUESTED** [review]
regression its doesn't work, can we add a tests please


regession

**Comment by jhenderson** [comment]
@defong Pushed fix

**Comment by defong** [comment]
Validation message?

image

image


2000+ years ago?

image

File: app/views/trainees/start_dates/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<%= render PageTitle::View.new(title: "trainees.trainee_ids.edit") %>
+
+<%= content_for(:breadcrumbs) do %>
+  <%= render DynamicBackLink::View.new(@trainee, text: t(:back_to_record)) %>
+<% end %>
+
+<%= form_with model: @training_details, url: trainee_start_date_path, local: true do |f| %>
+  <%= f.govuk_date_field :commencement_date, legend: {
+    text: t("views.forms.training_details.commencement_date.label"), size: 's'
+  }, hint: { text: t("views.forms.training_details.commencement_date.hint") + " 9 2 2021" } %>
```
Comment (edwardhorsford) [inline app/views/trainees/start_dates/edit.html.erb:10]:
```suggestion
  }, hint: { text: t("views.forms.training_details.commencement_date.hint") + " 15 2 2021" } %>
```

File: config/locales/en.yml [hunk 2]
```diff
@@ -198,6 +204,15 @@ en:
         labels:
           additional_reason: Enter the reason for withdrawal
           <<: *withdrawal_reasons
+      training_details:
+        title: Trainee start date and ID
+        commencement_date:
+          label: Date trainee started
+          hint: When this trainee started, rather than the programme start date. For example,
```
Comment (edwardhorsford) [inline config/locales/en.yml:211]: Can we put a `<br>` between the two sentences?

File: app/views/trainees/start_dates/edit.html.erb (hunk 1 above)
Comment (edwardhorsford) [reply app/views/trainees/start_dates/edit.html.erb:10]: What's the benefit of having the hint stored in two places?

**Review by edwardhorsford — CHANGES_REQUESTED** [review]
Looking good. A few copy things to fix.

##Heading on draft task
Current:
Screenshot 2021-02-15 at 11 03 31
Should be: `Confirm trainee start date and ID

## Edit page for start date missing heading
Screenshot 2021-02-15 at 11 04 27

The legend here doesn't appear to be marked up as a heading - I think we need the `isPageHeading` option or similar.

## Text wrong on start date confirmation page
Heading and summary card label slightly different
Screenshot 2021-02-15 at 11 04 54

## Wrong pluralisation on flash message
Screenshot 2021-02-15 at 11 05 04

## Need default label for records that don't have a start date
Can we add default text of `Not provided` if we don't have a start date?
Screenshot 2021-02-15 at 11 04 06

We should probably update the seeds so that seed records have this.

File: app/views/trainees/start_dates/edit.html.erb (hunk 1 above)
Comment (jhenderson) [reply app/views/trainees/start_dates/edit.html.erb:10]: It's only store in one place but used in 2 places, namely `app/views/trainees/start_dates/edit.html.erb` and `app/views/trainees/training_details/edit.html.erb`

Comment (edwardhorsford) [reply app/views/trainees/start_dates/edit.html.erb:10]: @jhenderson no I meant that the hint is compiled from two sources - the date gets appended on.

File: app/forms/training_details_form.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  include ActiveModel::AttributeAssignment
+  include ActiveModel::Validations::Callbacks
+
+  attr_accessor :trainee, :trainee_id, :day, :month, :year, :validate_commencement_date, :validate_trainee_id
+
+  delegate :id, :persisted?, to: :trainee
+
+  validates :trainee_id, presence: true, if: -> { validate_trainee_id }
+
+  validate :commencement_date_valid, if: -> { validate_commencement_date }
```
Comment (defong) [inline app/forms/training_details_form.rb:14]: Did we come to a conclusion?

> 2000+ years ago?
> 
> image

Comment (jhenderson) [reply app/forms/training_details_form.rb:14]: @defong Nick said a valid date will suffice. I guess we can create another card later on if necessary.

Comment (edwardhorsford) [reply app/forms/training_details_form.rb:14]: I'd probably match the programme start date validation. If not for this card, then we can raise another card.

File: config/locales/en.yml [hunk 4]
```diff
@@ -216,6 +231,10 @@ en:
           qts_awarded: QTS awarded
     errors:
       models:
+        trainee:
+          attributes:
+            record_type: You must select a route
+            trainee_id: Your entry must not exceed 100 characters
```
Comment (defong) [inline config/locales/en.yml:237]: 😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂😂

Trapped 
image

No validation message....

File: spec/factories/trainees.rb [hunk 5]
```diff
@@ -51,6 +51,10 @@
       programme_end_date { Faker::Date.between(from: programme_start_date + 1.day, to: Time.zone.today) }
     end
 
+    trait :with_start_date do
+      commencement_date { 6.months.from_now }
```
Comment (defong) [inline spec/factories/trainees.rb:55]:
```suggestion
      commencement_date { Faker::Date.between(from: 6.months.from_now, to: Time.zone.today) }
```

File: spec/features/trainees/creating_a_new_trainee_spec.rb [hunk 6]
```diff
@@ -3,17 +3,23 @@
...
     and_i_select_assessment_only_route
-    and_i_fill_in_train_id
     and_i_save_the_form
     then_i_should_see_the_new_trainee_overview
   end
 
+  scenario "submitting without choosing a route" do
+    and_i_save_the_form
+    then_i_should_see_a_validation_error
+  end
```
Comment (defong) [inline spec/features/trainees/creating_a_new_trainee_spec.rb:21]: :+1:

**Review by defong — APPROVED** [review]
Looks good to me...

Some minor bits, no showstoppper

**Comment by edwardhorsford** [comment]
I think we need some better validation on dates. Filling in the day and month *only* is accepted as valid - it appears to convert the empty year in to year 0, which is then considered valid.

Screenshot 2021-02-16 at 14 04 32

**Comment by edwardhorsford** [comment]
The summary card for trainee start date and id should exist on the 'review this record' page on draft records. At the moment it's showing the old summary card.

Screenshot 2021-02-16 at 14 07 30

File: config/locales/en.yml (hunk 4 above)
Comment (jhenderson) [reply config/locales/en.yml]: Good spot. Fixed.

**Comment by jhenderson** [comment]
@edwardhorsford Pushed more updates. Hopefully, that should be all the issues sorted.

**Review by edwardhorsford — APPROVED** [review]

**Pull request merged**